### PR TITLE
handle no_std targets on std builds

### LINKED ISSUE #128027: `./x check std` should not try to check `no_std` targets
`no_std` targets don't support std so checking is pointless and can be skipped. Or at the very least it could inform users that failing checks is expected and not an error.

Maybe there could be some way to force a check if someone really wants to.

### PR DESCRIPTION
**Pull request by onur-ozkan**
This PR unifies the `Step::run_make` logic and improves it by skipping std specific crates for no_std targets. In addition, since we now handle library crates properly, bootstrap is capable of running `x doc library` even for no_std targets as it is able to generate documentation for `alloc` crate from the standard library.

Resolves #128027

cc @ChrisDenton

### CODE REVIEW

**Comment by ChrisDenton** [comment]
This seems to work locally on individual but I am getting ICEs when checking a lot of targets at once. E.g. `./x check std aarch64-kmc-solid_asp3` works on its own but when included with a large number of other targets I get:

```
thread 'rustc' panicked at compiler\rustc_codegen_ssa\src\back\link.rs:86:13:
compiler\rustc_codegen_ssa\src\back\link.rs:86:13: invalid output type `Executable` for target os `aarch64-kmc-solid_asp3`
```

I've not managed to narrowed it down but it could be there's a specific other target that's causing issues when they're used together.

**Comment by onur-ozkan** [comment]
> it could be there's a specific other target that's causing issues when they're used together.

Probably.

File: src/bootstrap/src/core/build_steps/compile.rs [hunk 1]
```diff
@@ -424,6 +420,28 @@ fn copy_self_contained_objects(
     target_deps
 }
 
+/// Resolves standard library crates for `Std::run_make` for any build kind (like check, build, clippy, etc.).
+pub fn std_crates_for_run_make(run: &RunConfig<'_>) -> Vec<String> {
+    // FIXME: Extend builder tests to cover the `crates` field of `Std` instances.
+    if cfg!(feature = "bootstrap-self-test") {
+        return vec![];
+    }
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/compile.rs:428]: Additional context:

https://github.com/rust-lang/rust/blob/72d73cec61aa8f85901358cd5d386d5dd066fe52/src/bootstrap/src/core/builder/tests.rs#L60-L62

That above and `doc_std` helper macros needs to be improved to mock `crates` field in `Std` with relevant values.

**Comment by onur-ozkan** [comment]
Rebased.

@bors r=Mark-Simulacrum

**Pull request merged**